docs/designs/graceful-orphan-deletion.md

```diff
@@ -0,0 +1,195 @@
+
+The updated flow within `doReconcile` will be as follows:
+
+1.  The `doReconcile` function calls `model.AdapterForObject`.
+2.  Inside `AdapterForObject`:
+    *   It checks for a deletion timestamp.
+    *   If found, it checks for `status.externalRef`.
+    *   If `externalRef` exists, it returns a specialized adapter that uses it.
+    *   If not, it proceeds to resolve dependencies from the `spec` as usual.
+3.  Back in `doReconcile`:
+    *   It checks the error returned from `AdapterForObject`.
+    *   **If the error is `DependencyNotFound` AND the resource has a deletion timestamp**, it triggers the new orphan-handling logic: log a warning and call `handleDeleted()` (which removes the finalizer).
+    *   If the error is `DependencyNotFound` during a create/update, it preserves the existing behavior of retrying.
+    *   If no error occurs, it proceeds with the `Adapter` it received. The subsequent `adapter.Delete()` call will succeed whether the adapter was built from `status` or `spec`.
+
+### 5.2. Pseudocode for Implementation
+
+This pseudocode reflects the necessary changes in the resource-specific model and the generic `directbase` controller.
+
+**1. Conceptual Change in `(model *Model) AdapterForObject(...)`**
+*(This logic would be implemented in each direct controller's model, e.g., `gkehub/featuremembership_model.go`)*
+
+```go
+// In the model implementation for a resource...
+func (m *model) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+    // If the resource is being deleted AND has a status.externalRef,
+    // construct a specialized adapter that can perform Find/Delete using only that identifier.
+    // This bypasses spec dependency resolution entirely.
+    if !u.GetDeletionTimestamp().IsZero() {
+        externalRef, found, _ := unstructured.NestedString(u.Object, "status", "externalRef")
+        if found && externalRef != "" {
+            return newAdapterFromExternalRef(u, externalRef) // This adapter does not need resolved dependencies.
+        }
+    }
+
+    // For create, update, or deletion where status.externalRef is missing,
+    // proceed with the original logic of resolving dependencies from the spec.
+    // This will correctly return an UnresolvableDeps error if parents are missing.
+    if err := m.resolveDependencies(u); err != nil {
+        return nil, err
+    }
+    return newAdapterFromSpec(u, m.resolvedRefs)
+}
+```
+
+**2. Change in `(r *reconcileContext) doReconcile(...)`**
+*(Located in `pkg/controller/direct/directbase/directbase_controller.go`)*
+
+```go
+func (r *reconcileContext) doReconcile(ctx context.Context, u *unstructured.Unstructured) (requeue bool, err error) {
+    // ... (actuation mode check, defaulters) ...
+
+    // Get the adapter. The model's AdapterForObject now contains the deletion-specific logic.
+	adapter, adapterErr := r.Reconciler.model.AdapterForObject(ctx, r.Reconciler.Client, u)
```

Comment:
We can always iterate on the code, I typically prefer to have code examples in code